File: src/main/java/backend/model/user/UserRole.java
package backend.model.user;

import org.springframework.security.core.GrantedAuthority;

import javax.persistence.*;

@Entity
@Inheritance
@DiscriminatorColumn(name = "ROLE_NAME")
public abstract class UserRole implements User, GrantedAuthority {

    @Id
    @GeneratedValue()
    private Integer roleID;

    @ManyToOne
    private UserCore core;

    public UserRole() {

    }

    public UserRole(UserCore core) {
        this.core = core;
    }

    public static UserRole createFor(Class<? extends UserRole> clazz, UserCore core) throws Exception {
        return clazz.asSubclass(clazz).getConstructor(UserCore.class).newInstance(core);
    }

    @Override
    public UserCore getCore() {
        return this.core;
    }

    public void setCore(UserCore core) {
        this.core = core;
    }

    public Integer getRoleID() {
        return roleID;
    }

    public void setRoleID(Integer id) {
        this.roleID = id;
    }

    @Override
    public String getFirstname() {
        return this.core.getFirstname();
    }

    @Override
    public void setFirstname(String firstname) {
        this.core.setFirstname(firstname);
    }

    @Override
    public String getLastname() {
        return this.core.getLastname();
    }

    @Override
    public void setLastname(String lastname) {
        this.core.setLastname(lastname);
    }

    @Override
    public String getEmail() {
        return this.core.getEmail();
    }

    @Override
    public void setEmail(String email) {
        this.core.setEmail(email);
    }

    @Override
    public boolean isBlocked() {
        return this.core.isBlocked();
    }

    @Override
    public void setIsBlocked(boolean isBlocked) {
        this.core.setIsBlocked(isBlocked);
    }

    @Override
    public String getPassword() {
        return this.core.getPassword();
    }

    @Override
    public void setPassword(String password) {
        this.core.setPassword(password);
    }

    @Override
    public String getGender() {
        return this.core.getGender();
    }

    @Override
    public void setGender(String gender) {
        this.core.setGender(gender);
    }

    @Override
    public UserRole addRole(Class<? extends UserRole> clazz) throws Exception {
        return this.core.addRole(clazz);
    }

    @Override
    public UserRole getRole(Class<? extends UserRole> clazz) {
        return this.core.getRole(clazz);
    }

    @Override
    public boolean hasRole(Class<? extends UserRole> clazz) {
        return this.core.hasRole(clazz);
    }

    @Override
    public UserRole removeRole(Class<? extends UserRole> clazz) {
        return this.core.removeRole(clazz);
    }
}

File: src/main/java/backend/model/user/UserRole.kt
package backend.model.user

import backend.model.BasicEntity
import backend.model.media.Media
import com.fasterxml.jackson.annotation.JsonIgnore
import org.springframework.security.core.GrantedAuthority
import javax.persistence.DiscriminatorColumn
import javax.persistence.Entity
import javax.persistence.Inheritance
import javax.persistence.ManyToOne
import kotlin.reflect.KClass

@Entity
@Inheritance
@DiscriminatorColumn(name = "ROLE_NAME")
abstract class UserRole : BasicEntity, User, GrantedAuthority {

    @ManyToOne
    @JsonIgnore
    override lateinit var account: UserAccount

    /**
     * No args constructor needed for
     * private no args constructor of subclasses used for JPA
     */
    constructor() : super()

    constructor(account: UserAccount) : this() {
        this.account = account
    }

    // From here on: delegation of methods to account

    override var firstname: String?
        get() = this.account.firstname
        set(firstname) {
            this.account.firstname = firstname
        }

    override var lastname: String?
        get() = this.account.lastname
        set(lastname) {
            this.account.lastname = lastname
        }

    override var email: String
        get() = this.account.email
        set(email) {
            this.account.email = email
        }

    override var isBlocked: Boolean
        get() = this.account.isBlocked
        set(isBlocked) {
            this.account.isBlocked = isBlocked
        }

    override var passwordHash: String
        get() = this.account.passwordHash
        set(password) {
            this.account.passwordHash = password
        }

    override var gender: String?
        get() = this.account.gender
        set(gender) {
            this.account.gender = gender
        }

    override var profilePic: Media?
        get() = this.account.profilePic
        set(profilePic) {
            this.account.profilePic = profilePic
        }

    override var notificationToken: String?
        get() = this.account.notificationToken
        set(notificationToken) {
            this.account.notificationToken = notificationToken
        }

    override var preferredLanguage: Language
        get() = this.account.preferredLanguage
        set(value) {
            this.account.preferredLanguage = value
        }


    override fun emailDomain(): String {
        return email.split("@").last()
    }

    override fun <T : UserRole> addRole(clazz: KClass<T>): T = this.account.addRole(clazz)
    override fun <T : UserRole> getRole(clazz: KClass<T>): T? = this.account.getRole(clazz)
    override fun <T : UserRole> hasRole(clazz: KClass<T>): Boolean = this.account.hasRole(clazz)
    override fun <T : UserRole> removeRole(clazz: KClass<T>): T? = this.account.removeRole(clazz)

    override fun activate(token: String) = this.account.activate(token)
    override fun isActivationTokenCorrect(token: String): Boolean = this.account.isActivationTokenCorrect(token)
    override fun createActivationToken(): String = this.account.createActivationToken()
    override fun isActivated(): Boolean = this.account.isActivated()
    override fun setNewPassword(password: String, token: String) = this.account.setNewPassword(password, token)

    companion object {

        /**
         * Create new instance of a subclass of userRole
         * @param clazz The class of which an instance should be created. Must be subclass of UserRole
         *              and have a noargs constructor
         * @param account  The account with the UserRole will have
         * @return Returns a new instance of the specified class with a corresponding UserAccount
         */
        @Suppress("UNCHECKED_CAST")
        @Throws(Exception::class)
        fun <T : UserRole> createFor(clazz: Class<T>, account: UserAccount): UserRole {
            val constructor = clazz.declaredConstructors.firstOrNull { it.parameterCount == 0 }
                    ?: throw Exception("no args constructor not found on $clazz")
            constructor.isAccessible = true
            val o = constructor.newInstance() as T
            o.account = account
            return o
        }
    }
}
